lib/spack/spack/cmd/module.py

```diff
@@ -23,135 +23,206 @@
+    type_help = 'Type of module files'
+    subparser.add_argument('--module-type', help=type_help, default='tcl', choices=module_types)
+    constraint_help = 'Optional constraint to select a subset of installed packages'
+    subparser.add_argument('constraint', nargs='*', help=constraint_help)
 
-    find_parser = sp.add_parser('find', help='Find module files for packages.')
 
-    find_parser.add_argument(
-        'module_type',
-        help="Type of module to find file for. [" +
-        '|'.join(module_types) + "]")
+def setup_parser(subparser):
+    sp = subparser.add_subparsers(metavar='SUBCOMMAND', dest='module_command')
+    # spack module refresh
+    refresh_parser = sp.add_parser('refresh', help='Regenerate all module files.')
+    refresh_parser.add_argument('--delete-tree', help='Delete the module file tree before refresh', action='store_true')
+    _add_common_arguments(refresh_parser)
 
-    find_parser.add_argument(
+    # spack module find
+    find_parser = sp.add_parser('find', help='Find module files for packages.')
+    _add_common_arguments(find_parser)
+
+    # spack module env
+    loadlist_parser = sp.add_parser(
+        'load-list',
+        help='Prompt the list of modules associated with packages that satisfy a contraint'
+    )
+    loadlist_parser.add_argument(
         '-r', '--dependencies', action='store_true',
         dest='recurse_dependencies',
         help='Recursively traverse dependencies for modules to load.')
 
-    find_parser.add_argument(
+    loadlist_parser.add_argument(
         '-s', '--shell', action='store_true', dest='shell',
         help='Generate shell script (instead of input for module command)')
 
-    find_parser.add_argument(
-        '-p', '--prefix', dest='prefix',
+    loadlist_parser.add_argument(
+        '-p', '--prefix', dest='prefix', default='',
         help='Prepend to module names when issuing module load commands')
+    _add_common_arguments(loadlist_parser)
+
+
+class MultipleMatches(Exception):
+    pass
+
+
+class NoMatch(Exception):
+    pass
+
 
-    find_parser.add_argument(
-        'spec', nargs='+',
-        help='spec to find a module file for.')
+def load_list(mtype, specs, args):
 
+    # Get a comprehensive list of specs
+    if args.recurse_dependencies:
+        specs_from_user_constraint = specs[:]
+        specs = []
+        # FIXME : during module file creation nodes seem to be visited multiple
+        # FIXME : times even if cover='nodes' is given. This work around permits
+        # FIXME : to get a unique list of spec anyhow. Do we miss a merge
+        # FIXME : step among nodes that refer to the same package?
+        # FIXME : (same problem as in spack/modules.py)
+        seen = set()
+        seen_add = seen.add
+        for spec in specs_from_user_constraint:
+            specs.extend(
+                [item for item in spec.traverse(order='post', cover='nodes') if not (item in seen or seen_add(item))]
```

Comment:
Ah, nope. `traverse` is right. I think the comment probably only applies to concretized specs, otherwise the DAG does not combine its duplicate specs (by name).